doc/CFG.md

```diff
@@ -0,0 +1,108 @@
+
+CFG is built for following kinds of trees: SCRIPT, FUNCTION_DECLARATION, FUNCTION_EXPRESSION, METHOD_DECLARATION.
+
+### For Loop
+The `for` statement creates multiple blocks:
+
+- one for initialization of variables (before body)
+- one for the body of the for
+- one for update (following the body)
+- one for the condition
+
+### Switch
+
+The SWITCH statement is modelled as multiple `if-elseif-else` blocks. Interesting feature of `switch` statement in PHP is the fact that `continue` inside switch behaves the same as `break` statement ([see manual](http://php.net/manual/en/control-structures.continue.php))
+
+### Try Catch Finally
+
+We made several assumptions to simplify construction of CFG for `try-catch-finally` construct
+
+* To limit number of blocks and edges in the graph whole `try` body is treated as single block with `catch` and `finally` treated as successors of this block, while in fact every statement in the body could raise the exception and exit the block. It is up to the analysis using the CFG to take into consideration that body may not be executed as a whole.
+* If the `finally` clause is missing we assume as if empty `finally` clause was present to simplify the construction of CFG. 
+* We always assume that it is possible that exception is not being handled by any present `catch` clause, even if there is `catch` clause handling all `Exception` types. In particular this means that there is always path in the CFG representing exception escaping the function scope and reaching `END` block (visiting all enclosed `finally` blocks on the way)
+
+### Limitations
+
+#### Empty blocks
+
+During construction of CFG it might happen that some empty blocks are created (e.g. empty `finally` clause, ...). We have a mechanism to cleanup these empty blocks, however it is limited that it can only remove empty blocks with single predecessor and successor. If empty block has multiple successors (which happens often with `try-catch-finally` statements) it is not removed.
+
+#### Impossible paths from `finally`
+
+Because we are building using bottom-up approach, we will construct `finally` block before we construct the block for the `try` body. This `finally` block will have two successors:
+
+1. "exceptional" successor representing the path for the unhandled exception connecting it to the outer `finally` clause or  `END` block.
+2.  "normal" successor to the statement following the `try` statement.
+
+This "normal" successor can be infeasible in case when there is abrupt termination of the control flow like using `return` or `break` in the try body. Consider the following example
```

Comment:
```suggestion
This "normal" successor can be infeasible in case when there is an abrupt termination of the control flow like using `return` or `break` in the try body. Consider the following example
```